File: aoc_2023/day_23/part_2.py
from collections import defaultdict

from utils.file import read_input

contents = read_input(__file__)
# contents = """#.#####################
# #.......#########...###
# #######.#########.#.###
# ###.....#.>.>.###.#.###
# ###v#####.#v#.###.#.###
# ###.>...#.#.#.....#...#
# ###v###.#.#.#########.#
# ###...#.#.#.......#...#
# #####.#.#.#######.#.###
# #.....#.#.#.......#...#
# #.#####.#.#.#########v#
# #.#...#...#...###...>.#
# #.#.#v#######v###.###v#
# #...#.>.#...>.>.#.###.#
# #####v#.#.###v#.#.###.#
# #.....#...#...#.#.#...#
# #.#########.###.#.#.###
# #...###...#...#...#.###
# ###.###.#.###v#####v###
# #...#...#.#.>.>.#.>.###
# #.###.###.#.###.#.#v###
# #.....###...###...#...#
# #####################.#"""

grid = [line for line in contents.splitlines()]
rs, cs = (0, 0)
for c, char in enumerate(grid[0]):
    if char == ".":
        cs = c
        break
re, ce = (len(grid) - 1, 0)
for c, char in enumerate(grid[-1]):
    if char == ".":
        ce = c
        break

dirs = [(0, 1), (1, 0), (-1, 0), (0, -1)]

type Point = tuple[int, int]

edges: dict[Point, list[tuple[Point, int]]] = defaultdict(list)

max_dist = 0
seen: set[Point] = {(rs, cs)}


def get_points(point: Point) -> list[Point]:
    r, c = point
    valid_points: list[Point] = []
    for rd, cd in dirs:
        nr, nc = (r + rd, c + cd)
        if not (0 <= nr < len(grid) and 0 <= nc < len(grid[0])):
            continue
        if grid[nr][nc] == "#":
            continue
        valid_points.append((nr, nc))
    return valid_points


def traverse(start_point: Point):
    points = get_points(start_point)
    for point in points:
        if point in seen:
            continue
        seen.add(point)
        traverse_path(start_point, point)


def traverse_path(start_point: Point, point: Point):
    prev_point = start_point
    current_point = point
    dist = 0
    while True:
        dist += 1
        points = [p for p in get_points(current_point) if p != prev_point]
        if len(points) != 1:
            edges[current_point].append((start_point, dist))
            edges[start_point].append((current_point, dist))
            if current_point not in seen:
                seen.add(current_point)
                if len(points) > 0:
                    traverse(current_point)
            return
        seen.add(current_point)
        prev_point = current_point
        current_point = points[0]


traverse((rs, cs))


seen: set[Point] = set()


def longest():
    # 140s run time
    paths: list[tuple[Point, int, set[Point]]] = [((rs, cs), 0, {(rs, cs)})]
    max_dist = 0
    counter = 0
    while len(paths):
        counter += 1
        next_paths: list[tuple[Point, int, set[Point]]] = []
        for point, dist, seen in paths:
            for point, next_dist in edges[point]:
                total_dist = dist + next_dist
                if point == (re, ce):
                    max_dist = max(max_dist, total_dist)
                    continue
                if point in seen:
                    continue
                next_seen = seen.copy()
                next_seen.add(point)
                next_paths.append((point, total_dist, next_seen))
        paths = next_paths
    return max_dist


def longest_recur(point: Point, dist: int) -> int:
    # 12s run time
    if point == (re, ce):
        return dist
    max_dist = 0
    seen.add(point)
    for next_point, next_dist in edges[point]:
        if next_point in seen:
            continue
        l = longest_recur(next_point, dist + next_dist)
        max_dist = max(l, max_dist)
    seen.remove(point)
    return max_dist


edges_arr: list[list[tuple[int, int]]] = []
edge_to_arr: dict[Point, int] = {}


def ensure_default_edge(p: Point):
    if not p in edge_to_arr:
        edge_to_arr[p] = len(edges_arr)
        edges_arr.append([])


for k, v in edges.items():
    ensure_default_edge(k)
    edge_idx = edge_to_arr[k]
    for point, dist in v:
        ensure_default_edge(point)
        point_idx = edge_to_arr[point]
        edges_arr[edge_idx].append((point_idx, dist))

start_idx = edge_to_arr[(rs, cs)]
end_idx = edge_to_arr[(re, ce)]
seen_idx: set[int] = set()


def longest_arr_recur(point: int, dist: int) -> int:
    # 9.8s run time, slightly faster to use idx's over dict but not that much
    if point == end_idx:
        return dist
    max_dist = 0
    seen_idx.add(point)
    for next_point, next_dist in edges_arr[point]:
        if next_point in seen_idx:
            continue
        l = longest_arr_recur(next_point, dist + next_dist)
        max_dist = max(l, max_dist)
    seen_idx.remove(point)
    return max_dist


# print(longest())
# print(longest_recur((rs, cs), 0))
print(longest_arr_recur(start_idx, 0))

File: aoc_2023/day_23/part_2.rs
use std::collections::HashMap;
use std::collections::HashSet;

#[path = "../../utils/file.rs"]
mod file;

fn main() {
    let contents = file::read_file(file_abs!());
    //     let contents = "#.#####################
    // #.......#########...###
    // #######.#########.#.###
    // ###.....#.>.>.###.#.###
    // ###v#####.#v#.###.#.###
    // ###.>...#.#.#.....#...#
    // ###v###.#.#.#########.#
    // ###...#.#.#.......#...#
    // #####.#.#.#######.#.###
    // #.....#.#.#.......#...#
    // #.#####.#.#.#########v#
    // #.#...#...#...###...>.#
    // #.#.#v#######v###.###v#
    // #...#.>.#...>.>.#.###.#
    // #####v#.#.###v#.#.###.#
    // #.....#...#...#.#.#...#
    // #.#########.###.#.#.###
    // #...###...#...#...#.###
    // ###.###.#.###v#####v###
    // #...#...#.#.>.>.#.>.###
    // #.###.###.#.###.#.#v###
    // #.....###...###...#...#
    // #####################.#";

    let grid = contents
        .split("\n")
        .map(|line| line.chars().collect::<Vec<_>>())
        .collect::<Vec<_>>();

    let sc = grid[0].iter().position(|&x| x == '.').unwrap();
    let start = (0, sc);
    let ec = grid.last().unwrap().iter().position(|&x| x == '.').unwrap();
    let end = (grid.len() - 1, ec);

    let mut graph: Graph = HashMap::new();
    let mut seen: HashSet<(usize, usize)> = HashSet::new();
    println!("traversing for graph...");
    traverse_graph(start, &grid, &mut graph, &mut seen);

    println!("running longest path algo...");
    if false {
        // takes around 6s
        let mut seen: HashSet<(usize, usize)> = HashSet::new();
        let result = longest_recur(start, &graph, &mut seen, end, 0);
        println!("Result: {}", result);
    }

    println!("using no hashing...");
    let (mut vec_graph, tuple_to_idx) = hash_to_vec_graph(&graph);
    let start_idx = tuple_to_idx[&start];
    let end_idx = tuple_to_idx[&end];

    let mut seen_vec = vec![false; vec_graph.len()];
    // let mut counter = 0;
    let result = longest_recur_vec(
        start_idx,
        &mut vec_graph,
        &mut seen_vec,
        end_idx,
        // &mut counter,
        0,
    );
    // println!("counter: {:?}", counter);
    println!("Result: {}", result)
}

type VecGraph = Vec<Vec<(usize, usize)>>;
fn hash_to_vec_graph(graph: &Graph) -> (VecGraph, HashMap<(usize, usize), usize>) {
    let mut vec_graph: VecGraph = Vec::new();
    let mut tuple_to_idx: HashMap<(usize, usize), usize> = HashMap::new();
    let mut get_with_default = |x: &(usize, usize), v: &mut Vec<Vec<(usize, usize)>>| -> usize {
        tuple_to_idx.get(x).copied().unwrap_or_else(|| {
            let result = v.len();
            tuple_to_idx.insert(*x, result);
            v.push(Vec::new());
            return result;
        })
    };
    for (k, v) in graph {
        let k_idx = get_with_default(k, &mut vec_graph);
        for (item, dist) in v {
            let edge_idx = get_with_default(item, &mut vec_graph);
            vec_graph[k_idx].push((edge_idx, *dist));
        }
    }
    return (vec_graph, tuple_to_idx);
}

type Graph = HashMap<(usize, usize), Vec<((usize, usize), usize)>>;
fn traverse_graph(
    start_point: (usize, usize),
    grid: &Vec<Vec<char>>,
    graph: &mut Graph,
    seen: &mut HashSet<(usize, usize)>,
) {
    // get all possible routes, filter by not previous
    // if routes.count > 1, done
    // go to next
    let points = get_points(start_point, grid);
    for point in points {
        if seen.contains(&point) {
            continue;
        }
        seen.insert(point);
        traverse_path(start_point, point, grid, graph, seen);
    }
}
fn traverse_path(
    start_point: (usize, usize),
    current_point: (usize, usize),
    grid: &Vec<Vec<char>>,
    graph: &mut Graph,
    seen: &mut HashSet<(usize, usize)>,
) {
    let mut prev_point = start_point;
    let mut last_point = current_point;
    let mut dist = 0;
    loop {
        dist += 1;
        let points = get_points(last_point, grid)
            .iter()
            .filter(|&p| *p != prev_point)
            .copied()
            .collect::<Vec<_>>();
        if points.len() != 1 {
            graph
                .entry(last_point)
                .or_insert(Vec::new())
                .push((start_point, dist));
            graph
                .entry(start_point)
                .or_insert(Vec::new())
                .push((last_point, dist));
            if !seen.contains(&last_point) {
                seen.insert(last_point);
                if points.len() > 1 {
                    traverse_graph(last_point, grid, graph, seen);
                }
            }
            return;
        }
        prev_point = last_point;
        seen.insert(last_point);
        last_point = *points.first().unwrap();
    }
}
fn get_points((r, c): (usize, usize), grid: &Vec<Vec<char>>) -> Vec<(usize, usize)> {
    let dirs: [(isize, isize); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];
    return dirs
        .iter()
        .filter_map(|(rd, cd)| {
            let (nr, nc) = (r.checked_add_signed(*rd), c.checked_add_signed(*cd));
            if let (Some(nr), Some(nc)) = (nr, nc) {
                if nr >= grid.len() || nc >= grid[0].len() {
                    return None;
                }
                if grid[nr][nc] == '#' {
                    return None;
                }
                return Some((nr, nc));
            }
            return None;
        })
        .collect::<Vec<_>>();
}

fn longest_recur(
    point: (usize, usize),
    graph: &Graph,
    seen: &mut HashSet<(usize, usize)>,
    end_point: (usize, usize),
    // counter: &mut usize,
    dist: usize,
) -> usize {
    // *counter += 1;
    if point == end_point {
        return dist;
    }
    let mut max = 0;
    seen.insert(point);
    for &(next_point, next_dist) in &graph[&point] {
        if seen.contains(&next_point) {
            continue;
        }
        let l = longest_recur(
            next_point,
            graph,
            seen,
            end_point,
            // counter,
            dist + next_dist,
        );
        max = max.max(l);
    }
    seen.remove(&point);
    max
}

fn longest_recur_vec(
    point: usize,
    graph: &VecGraph,
    seen: &mut Vec<bool>,
    end_point: usize,
    // counter: &mut usize,
    dist: usize,
) -> usize {
    // *counter += 1;
    if point == end_point {
        return dist;
    }
    let mut max = 0;
    seen[point] = true;
    for &(next_point, next_dist) in &graph[point] {
        if seen[next_point] {
            continue;
        }
        let l = longest_recur_vec(
            next_point,
            graph,
            seen,
            end_point,
            // counter,
            dist + next_dist,
        );
        max = max.max(l);
    }
    seen[point] = false;
    max
}
